### PR TITLE
[nav] Gate Search

### PR DESCRIPTION
**Pull request by qhdwight**
- Spider search
- Dual segment search
- Combines all turn and drive states into one for increased simplicity and readability
- PID did not divide by dt for d term (I'm not even convinced we need anything other than a P-loop anyways tbh)

### CODE REVIEW

**Comment by qhdwight** [comment]
Will clean up the merge conflicts...

Prolly have to rebase properly

**Comment by qhdwight** [comment]
https://user-images.githubusercontent.com/20666629/162606663-648b7359-df83-4cc9-805b-d0851e11596f.mp4

File: jetson/nav/environment.cpp [hunk 1]
```diff
@@ -12,10 +19,143 @@ Obstacle Environment::getObstacle() {
...
+}
+
+Vector2d Environment::getPostTwoRelative(Odometry cur) {
+    double bearing = degreeToRadian(calcBearing(cur, mPostTwo));
+    double distance = estimateNoneuclid(cur, mPostTwo);
+    return {distance * cos(bearing), distance * sin(bearing)};
+}
+
+bool Environment::areTargetFiltersReady() const {
+    return mLeftDistanceFilter.ready() && mRightDistanceFilter.ready() && mLeftBearingFilter.ready() && mRightBearingFilter.ready();
```
Comment (zwgold) [inline jetson/nav/environment.cpp:143]: Any reason why you aren't just calling `isLeftTargetFilterReady()` and `isRightTargetFilterReady()`?

File: jetson/nav/filter.hpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include <array>
+#include <numeric>
+#include <algorithm>
+
+template<typename T>
+class Filter {
```
Comment (zwgold) [inline jetson/nav/filter.hpp:8]: Just for my sake, this is to replace the caching we were doing before, or is this for a diff reason?

File: jetson/nav/obstacle_avoidance/simpleAvoidance.cpp [hunk 3]
```diff
@@ -19,61 +19,13 @@ SimpleAvoidance::~SimpleAvoidance() = default;
...
-        (obstacleBearing < 0 ? mLastObstacleAngle >= 0 : mLastObstacleAngle < 0)) {
-        obstacleBearing *= -1;
-    }
-
-    double desiredBearing = mod(rover->odometry().bearing_deg + obstacleBearing, 360);
-    mJustDetectedObstacle = true;
-    mLastObstacleAngle = obstacleBearing;
-    rover->turn(desiredBearing);
-    return rover->currentState();
+    throw std::runtime_error("implement me");
```
Comment (zwgold) [inline jetson/nav/obstacle_avoidance/simpleAvoidance.cpp:22]: lol nice

File: jetson/nav/search/diamondSearch.txt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+0.0 90.0
```
Comment (zwgold) [inline jetson/nav/search/diamondSearch.txt:1]: Can we delete this? It seems redundant to have, plus we're removing the "DiamondSearch" we have in this PR

File: jetson/nav/search/searchStateMachine.cpp [hunk 5]
```diff
@@ -10,217 +10,166 @@
...
-        Obstacle obstacle = env->getObstacle();
-        sm->updateObstacleElements(obstacle.bearing, obstacle.rightBearing, obstacle.distance);
-        return NavState::SearchTurnAroundObs;
+    double distance, bearing;
+    if (lastWaypoint.gate) {
+        if (env->hasGateLocation()) {
+            return NavState::BeginGateSearch;
+        }
+        if (leftTarget.id >= 0 && rightTarget.id >= 0) {
+            if (leftTarget.distance < rightTarget.distance) {
```
Comment (zwgold) [inline jetson/nav/search/searchStateMachine.cpp:95]: This block from line 95 to about 100 is similar to how I had made it go to the closest post, but it resulted in the weird infinite loop between `TurnToTarget` and `DriveToTarget`. Did you test a case where we'd see a post that is supposed to go into the right side first (which would make it go into `leftTarget`) and then we'd see two posts, causing the one in the left to go to the right?

**Comment by zwgold** [comment]
image

@qhdwight Did this get tested on jarvis? I ran into this issue when compiling

File: jetson/nav/stateMachine.cpp [hunk 6]
```diff
@@ -103,26 +100,18 @@ void StateMachine::run() {
...
-        case NavState::GateTurn:
-        case NavState::GateDrive:
-        case NavState::GateTurnToCentPoint:
-        case NavState::GateDriveToCentPoint:
-        case NavState::GateFace:
-        case NavState::GateDriveThrough:
-        case NavState::GateTurnToFarPost:
-        case NavState::GateDriveToFarPost:
-        case NavState::GateTurnToGateCenter: {
+        case NavState::GateTraverse: {
```
Comment (zwgold) [inline jetson/nav/stateMachine.cpp:109]: Bit of duplicate code here for GateTraverse, called twice in switch

Comment (zwgold) [reply jetson/nav/stateMachine.cpp:109]: Also, seems that BeginGateSearch is not in here as one of the calls

File: jetson/nav/environment.cpp [hunk 7]
```diff
@@ -1,8 +1,15 @@
...
-
+#include "cmath"
 #include "environment.hpp"
+#include <iostream>
+#include <string>
 
-Environment::Environment() = default;
+Environment::Environment(const rapidjson::Document& config) :
+        mConfig(config),
+        mLeftBearingFilter(config["gate"]["filterSize"].GetInt(), config["gate"]["filterProportion"].GetDouble()),
```
Comment (zwgold) [inline jetson/nav/environment.cpp:9]: `filterProportion` not defined in the config, but `filterSize` is

**Review by cgiger00 — APPROVED** [review]

**Comment by cgiger00** [comment]
PRing in for testing

**Pull request merged**